### PR TITLE
Implement JSON <-> CBOR helper

### PR DESCRIPTION
**Pull request by GregBurns**
DPS is payload agnostic but uses CBOR encoding for the CBOR protocol itself.
This change provides a pair of APIs for translating between JSON and CBOR leveraging DPS's private CBOR APIs.
DPS_JSON2CBOR() translates a NULL terminated JSON string into a CBOR encode payload.
DPS_CBOR2JSON() translates a CBOR encoded payload into JSON, this API has a "pretty" option that formats the JSON for readability.
For encoding binary (byte string) data implementation borrows the MongoDB approach of using a map with a "$binary" key
followed by a hex encoded string.

### CODE REVIEW

File: examples/subscriber.c [hunk 1]
```diff
@@ -269,6 +279,11 @@ int main(int argc, char** argv)
                 DPS_Debug = 1;
                 continue;
             }
+            if (strcmp(*arg, "-j") == 0) {
```
Comment (malsbat) [inline examples/subscriber.c]: Add -j to Usage: section

File: inc/dps/json.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/**
+ * @file
+ * JSON <-> CBOR conversion
+ */
+
+/*
+ *******************************************************************
+ *
+ * Copyright 2016 Intel Corporation All rights reserved.
+ *
```
Comment (malsbat) [inline inc/dps/json.h]: 2018?

File: inc/dps/json.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Generate CBOR from a JSON string
+ *
+ * @param json     NUL terminated JSON string to convert
+ * @param cbor     Destination buffer for the conversion
+ * @param cborSize The size of the cbor buffer
+ * @param cborLen  Returns the number of CBOR bytes written 
+ *
+ * @return - DPS_OK if the conversion was successful
```
Comment (malsbat) [inline inc/dps/json.h]: Doxygen doesn't format the bullets correctly unless you do:
@return 
- bullet 1
- bullet 2

File: inc/dps/json.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ * @param json     JSON string to convert
+ * @param jsonSize The length of the JSON string buffer
+ * @param pretty   If TRUE format with indentation and newlines
+ *
+ * @return - DPS_OK if the conversion was successful
+ *         - DPS_ERR_OVERFLOW if the json buffer was too small
+ *         - DPS_ERR_INVALID if the input was not valid CBOR
+ *         - other error status codes
+ */
+DPS_Status DPS_CBOR2JSON(uint8_t* cbor, size_t cborLen, char* json, size_t jsonSize, int pretty);
```
Comment (malsbat) [inline inc/dps/json.h]: const uint8_t* cbor since it's an input param

File: src/cbor.c [hunk 5]
```diff
@@ -640,9 +635,6 @@ DPS_Status CBOR_DecodeDouble(DPS_RxBuffer* buffer, double* d)
     uint8_t* p = buffer->rxPos;
     uint8_t* pd;
 
-    if (avail < 9) {
-        return DPS_ERR_EOD;
-    }
```
Comment (malsbat) [inline src/cbor.c]: We still want if (avail < 1) { return DPS_ERR_EOD; } before derefing p

File: src/json.c [hunk 6]
```diff
@@ -0,0 +1,666 @@
...
+
+#include <ctype.h>
+#include <safe_lib.h>
+#include <string.h>
+#include <stdlib.h>
+#include <dps/private/cbor.h>
+#include <dps/dbg.h>
+#include <dps/json.h>
+
+#define MAX_INPUT_STRING_LEN  RSIZE_MAX_STR
```
Comment (malsbat) [inline src/json.c]: We may want to use (RSIZE_MAX_STR) - 1 here.  In all the other #define MAX cases we check if the value is <= RSIZE_MAX_STR.  Is there a test for values of input string around the max?

File: test/jsontest.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    int i;
+    // Check with and without formatting
+    for (pretty = 0; pretty <= 1; ++pretty) {
+        for (i = 0; i < (sizeof(tests) / sizeof(tests[0])); ++i) {
+            size_t cbor1Len;
+            size_t cbor2Len;
+
+            // Encode as CBOR
+            status = DPS_JSON2CBOR(tests[i], cbor1, sizeof(cbor1), &cbor1Len);
+            if (status != DPS_OK) {
```
Comment (malsbat) [inline test/jsontest.c]: replace if (status != DPS_OK) with ASSERT(status == DPS_OK) from test.h

File: test/jsontest.c [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    printf("Passed\n");
+    return EXIT_SUCCESS;
+
+Failed:
+
+    printf("Failed at test %d %s\n", i, DPS_ErrTxt(status));
+    return EXIT_FAILURE;
+}
```
Comment (malsbat) [inline test/jsontest.c]: it'd be good to have a json_fuzzer test too, similar to cbor_fuzzer, and also add json_test to test_scripts/run.py

File: inc/dps/json.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+ * @param cbor     Destination buffer for the conversion
+ * @param cborSize The size of the cbor buffer
+ * @param cborLen  Returns the number of CBOR bytes written 
+ *
+ * @return - DPS_OK if the conversion was successful
+ *         - DPS_ERR_OVERFLOW if the cbor buffer was too small
+ *         - DPS_ERR_INVALID if the input was not valid JSON
+ *         - other error status codes
+ */
+DPS_Status DPS_JSON2CBOR(const char* json, uint8_t* cbor, size_t cborSize, size_t* cborLen);
```
Comment (malsbat) [inline inc/dps/json.h]: new APIs will need to be added to dps_shared.def

File: inc/dps/json.h [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+ */
+DPS_Status DPS_CBOR2JSON(uint8_t* cbor, size_t cborLen, char* json, size_t jsonSize, int pretty);
+
+/** @} */
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
```
Comment (malsbat) [inline inc/dps/json.h]: did you want this to be part of the bindings?  the new header will need to be added to dps.i and renames to {js,py}/dps_map.i (the regexes I used to make the names more idiomatic don't like all caps).  

Also the typemaps for byte arrays require the name to be included unfortunately (see the code starting around line 382 in dps.i).

File: examples/subscriber.c (hunk 1 above)
Comment (GregBurns) [reply examples/subscriber.c]: Will fix

File: inc/dps/json.h (hunk 2 above)
Comment (GregBurns) [reply inc/dps/json.h]: Correct, will fix

File: inc/dps/json.h (hunk 4 above)
Comment (GregBurns) [reply inc/dps/json.h]: Will fix but even though the buffer is never modified the const needs to be cast away explicitly inside the function.

File: src/json.c (hunk 6 above)
Comment (GregBurns) [reply src/json.c]: Will fix

File: test/jsontest.c (hunk 7 above)
Comment (GregBurns) [reply test/jsontest.c]: Will change to use CHECK macro as used in cbortest.c

File: test/jsontest.c (hunk 8 above)
Comment (GregBurns) [reply test/jsontest.c]: Will add to test_scripts/run 
I agree a json_fuzzer test would be good. Will look into that later.

File: inc/dps/json.h (hunk 3 above)
Comment (GregBurns) [reply inc/dps/json.h]: Will fix

**Comment by malsbat** [comment]
Looks like most everything is fixed up, except:
- new APIs in dps_shared.def
- bindings
- json_fuzzer
I'd say dps_shared.def needs to be fixed up, but the others can wait.  I can help out with the bindings and fuzzer.

**Comment by GregBurns** [comment]
Fixed dps_shared.def.
I’ll leave the bindings to you.
A JSON fuzzer is a nice-to-have but given that the JSON is not sent over the air I am inclined to put a low priority on it. I will extend the test cases to provide coverage for the buffer overflow checks.

File: inc/dps/json.h (hunk 10 above)
Comment (GregBurns) [reply inc/dps/json.h]: Yes later

**Pull request merged**